data/transactions/logic/eval.go

```diff
@@ -3240,14 +3220,14 @@ func (cx *EvalContext) getLatestTimestamp() (uint64, error) {
 }
 
 // getApplicationAddress memoizes app.Address() across a tx group's evaluation
-func (cx *EvalContext) getApplicationAddress(app basics.AppIndex) basics.Address {
+func (ep *EvalParams) getApplicationAddress(app basics.AppIndex) basics.Address {
 	/* Do not instantiate the cache here, that would mask a programming error.
 	   The cache must be instantiated at EvalParams construction time, so that
 	   proper sharing with inner EvalParams can work. */
```

Comment:
this comment is weird now?

Reply:
I think it is still correct, though maybe it needs to be more clear.  The concern is that, when an inner EP is constructed, it uses the cache from its parent as its own cache.  If we create the cache map lazily, it might still be `nil` at that time, then the inner might add instantiate a map/cache, but that map/cache _won't be shared_ with the parent. So the cache is less effective overall.  Instead, we should not be lazy - we should `make` the map at EP construction time so that it exists (even if it's empty) when the inner is made, and thus they are shared.

Unless your point is "won't the cache always be instantiated, even if done lazily, at the top-level, now that we are building `ep.available` at the start of the group?  The answer to that is yes, I think. But that also means putting the lazy instantiation here is pointless.

Reply:
The thought process was, now that this method is associated with the struct in the comment, why call out a construction invariant for that struct, surely lots of things need to be done at EP construction time that would otherwise lead to errors, and since you're ensuring the construction invariant at construction time (presumably), what's the concern?  this was before I saw that EvalParams is the parent of EvalContext as opposed to the compositional relationship which I figured was driving the comment, and that this change is actually reassociating the method with the parent type.  So it was sort of along the lines of "won't this cache always be instantiated".

Reply:
We often find ourselves looking for optimizations, notice that a map is being made in a constructor but rarely used, stop doing that, and then change code that was writing to the map to perform the allocation lazily. Like this, in the same file and structure:
```
// Initialize innerTxidCache if necessary
if cx.EvalParams.innerTxidCache == nil {
	cx.EvalParams.innerTxidCache = make(map[int]transactions.Txid)
}
```

This comment is intended to remind the reader not to do that.